File: tools/company-graph/normalization/gbizinfo.py
"""
gBizINFO API クライアント + 法人番号名寄せエンジン

マッチング戦略:
  優先①: 社名 + 郵便番号 の完全一致
  優先②: 社名 + 代表者名 の完全一致（将来拡張）
  フォールバック: 表記揺れ正規化後の部分一致 (difflib)

API ドキュメント: https://info.gbiz.go.jp/hojin/swagger-ui.html
認証: X-hojinInfo-api-token ヘッダー（環境変数 GBIZINFO_API_KEY）
"""
from __future__ import annotations

import logging
import os
import re
import time
import unicodedata
from difflib import SequenceMatcher
from functools import lru_cache

import requests

from models import CompanyNode, RawCompany

logger = logging.getLogger(__name__)

GBIZINFO_BASE = "https://info.gbiz.go.jp/hojin/v1/hojin"

# 法人格の正規化テーブル
_CORP_FORMS = {
    "（株）": "株式会社",
    "(株)": "株式会社",
    "㈱": "株式会社",
    "（有）": "有限会社",
    "(有)": "有限会社",
    "㈲": "有限会社",
    "（合）": "合同会社",
    "(合)": "合同会社",
    "（資）": "合資会社",
    "(資)": "合資会社",
    "（名）": "合名会社",
    "(名)": "合名会社",
}


def normalize_name(name: str) -> str:
    """社名の表記揺れを正規化する"""
    # Unicode NFKC 正規化（全角→半角 etc.）
    name = unicodedata.normalize("NFKC", name)
    # 略称展開
    for abbr, full in _CORP_FORMS.items():
        name = name.replace(abbr, full)
    # 余分な空白を除去
    name = re.sub(r"\s+", "", name)
    return name.strip()


def _similarity(a: str, b: str) -> float:
    return SequenceMatcher(None, a, b).ratio()


class GBizInfoClient:
    def __init__(self, api_key: str | None = None, rate_limit_sec: float = 1.0) -> None:
        self._api_key = api_key or os.environ.get("GBIZINFO_API_KEY", "")
        self._rate_limit = rate_limit_sec
        self._last_req: float = 0.0
        self._session = requests.Session()
        if self._api_key:
            self._session.headers["X-hojinInfo-api-token"] = self._api_key

    def _wait(self) -> None:
        elapsed = time.monotonic() - self._last_req
        if elapsed < self._rate_limit:
            time.sleep(self._rate_limit - elapsed)
        self._last_req = time.monotonic()

    @lru_cache(maxsize=512)
    def search(self, name: str, postal_code: str = "") -> list[dict]:
        """gBizINFO API を呼び出して候補を返す"""
        self._wait()
        params: dict = {"name": name, "limit": 10}
        if postal_code:
            params["postal_code"] = postal_code.replace("-", "")
        try:
            resp = self._session.get(GBIZINFO_BASE, params=params, timeout=10)
            if resp.status_code == 401:
                logger.warning("gBizINFO: API key is missing or invalid.")
                return []
            resp.raise_for_status()
            return resp.json().get("hojin-infos", [])
        except requests.RequestException as exc:
            logger.error("gBizINFO API error (%s): %s", name, exc)
            return []


class NameMatcher:
    """RawCompany → CompanyNode への名寄せを担う"""

    def __init__(self, client: GBizInfoClient, threshold: float = 0.75) -> None:
        self._client = client
        self._threshold = threshold

    def match(self, raw: RawCompany) -> CompanyNode | None:
        normalized = normalize_name(raw.raw_name)

        # ① 社名 + 郵便番号 で検索
        candidates = self._client.search(normalized, raw.postal_code)
        if not candidates and raw.postal_code:
            # 郵便番号なしで再検索
            candidates = self._client.search(normalized)

        if not candidates:
            logger.warning("gBizINFO: no candidates for '%s'", raw.raw_name)
            return self._fallback(raw)

        best, best_score = self._rank(normalized, candidates)
        needs_review = best_score < self._threshold

        if needs_review:
            logger.warning(
                "Low match score %.2f for '%s' → '%s'",
                best_score, raw.raw_name, best.get("name", ""),
            )

        return CompanyNode(
            corporate_number=best.get("corporate_number", ""),
            official_name=best.get("name", raw.raw_name),
            source_urls=[raw.source_url],
            business_category=best.get("business_summary", {}).get(
                "major_classification_name", ""
            ),
            address=best.get("location", raw.address),
            website=raw.website,
            capital=raw.capital,
            employees=raw.employees,
            match_score=best_score,
            needs_review=needs_review,
            related_companies_text=raw.related_companies_text,
            business_partners_text=raw.business_partners_text,
        )

    def _rank(self, query: str, candidates: list[dict]) -> tuple[dict, float]:
        scored = []
        for c in candidates:
            candidate_name = normalize_name(c.get("name", ""))
            score = _similarity(query, candidate_name)
            scored.append((c, score))
        scored.sort(key=lambda x: x[1], reverse=True)
        return scored[0]

    def _fallback(self, raw: RawCompany) -> CompanyNode:
        """API で見つからなかった場合のフォールバックノード"""
        return CompanyNode(
            corporate_number="UNKNOWN_" + normalize_name(raw.raw_name),
            official_name=raw.raw_name,
            source_urls=[raw.source_url],
            address=raw.address,
            website=raw.website,
            capital=raw.capital,
            employees=raw.employees,
            match_score=0.0,
            needs_review=True,
            related_companies_text=raw.related_companies_text,
            business_partners_text=raw.business_partners_text,
        )

File: Backend/internal/scraper/gbizinfo.go
package scraper

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"sync"
	"time"

	"golang.org/x/time/rate"
)

// GBizRecord is one entry from the gBizINFO API.
type GBizRecord struct {
	CorporateNumber  string `json:"corporate_number"`
	Name             string `json:"name"`
	PostalCode       string `json:"postal_code"`
	Location         string `json:"location"`
	BusinessSummary  struct {
		MajorClassificationName string `json:"major_classification_name"`
	} `json:"business_summary"`
	CompanyURL string `json:"company_url"`
}

// GBizClient calls the gBizINFO API and normalises company names.
type GBizClient struct {
	BaseURL string
	Token   string
	Client  *http.Client
	Limiter *rate.Limiter

	mu    sync.Mutex
	cache map[string][]GBizRecord
}

func NewGBizClient(baseURL, token string) *GBizClient {
	if baseURL == "" {
		baseURL = "https://info.gbiz.go.jp/hojin/v1/hojin"
	}
	return &GBizClient{
		BaseURL: baseURL,
		Token:   token,
		Client:  &http.Client{Timeout: 15 * time.Second},
		Limiter: NewLimiter(1 * time.Second),
		cache:   map[string][]GBizRecord{},
	}
}

// Search queries gBizINFO for a company name (with optional postal code).
func (c *GBizClient) Search(ctx context.Context, name, postalCode string) ([]GBizRecord, error) {
	cacheKey := name + "|" + postalCode
	c.mu.Lock()
	if cached, ok := c.cache[cacheKey]; ok {
		c.mu.Unlock()
		return cached, nil
	}
	c.mu.Unlock()

	if err := c.Limiter.Wait(ctx); err != nil {
		return nil, err
	}

	params := url.Values{"name": {name}, "limit": {"10"}}
	if postalCode != "" {
		params.Set("postal_code", postalCode)
	}
	reqURL := c.BaseURL + "?" + params.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	if c.Token != "" {
		req.Header.Set("X-hojinInfo-api-token", c.Token)
	}

	resp, err := c.Client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("gbizinfo request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized {
		return nil, fmt.Errorf("gbizinfo: 401 Unauthorized (check GBIZINFO_API_TOKEN)")
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("gbizinfo: HTTP %d", resp.StatusCode)
	}

	var payload struct {
		Records []GBizRecord `json:"hojin-infos"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, fmt.Errorf("gbizinfo decode: %w", err)
	}

	c.mu.Lock()
	c.cache[cacheKey] = payload.Records
	c.mu.Unlock()

	return payload.Records, nil
}

// Match finds the best gBizINFO record for a RawCompany.
// Returns (record, score, error). When no record is found or token is missing,
// a fallback CompanyNode with corporate_number = "UNKNOWN_{name}" is returned.
func (c *GBizClient) Match(ctx context.Context, raw *RawCompany, threshold float64) (*CompanyNode, error) {
	normalized := NormalizeName(raw.RawName)

	// Try with postal code first, then without
	var candidates []GBizRecord
	var err error
	if raw.PostalCode != "" {
		candidates, err = c.Search(ctx, normalized, raw.PostalCode)
	}
	if err != nil || len(candidates) == 0 {
		candidates, err = c.Search(ctx, normalized, "")
	}

	if err != nil || len(candidates) == 0 {
		return fallbackNode(raw), nil
	}

	// Score each candidate
	best, bestScore := candidates[0], 0.0
	for _, rec := range candidates {
		s := Similarity(normalized, NormalizeName(rec.Name))
		if s > bestScore {
			bestScore = s
			best = rec
		}
	}

	needsReview := bestScore < threshold
	return &CompanyNode{
		CorporateNumber:      best.CorporateNumber,
		OfficialName:         best.Name,
		SourceURLs:           []string{raw.SourceURL},
		BusinessCategory:     best.BusinessSummary.MajorClassificationName,
		Address:              best.Location,
		Website:              best.CompanyURL,
		Capital:              raw.Capital,
		Employees:            raw.Employees,
		MatchScore:           bestScore,
		NeedsReview:          needsReview,
		RelatedCompaniesText: raw.RelatedCompaniesText,
		BusinessPartnersText: raw.BusinessPartnersText,
	}, nil
}

// SearchByKeyword は企業名キーワードで gBizINFO を検索し、CompanyNode のリストを返す。
// スクレイピングに代わる公式データ取得手段として使用する。
// gBizINFO が直接返す公式データのため、類似度判定は行わず MatchScore=1.0 で登録する。
func (c *GBizClient) SearchByKeyword(ctx context.Context, keyword string, limit int) ([]*CompanyNode, []string, error) {
	if limit <= 0 {
		limit = 20
	}
	if err := c.Limiter.Wait(ctx); err != nil {
		return nil, nil, err
	}

	params := url.Values{"name": {keyword}, "limit": {fmt.Sprintf("%d", limit)}}
	reqURL := c.BaseURL + "?" + params.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, nil, err
	}
	req.Header.Set("Accept", "application/json")
	if c.Token != "" {
		req.Header.Set("X-hojinInfo-api-token", c.Token)
	}

	resp, err := c.Client.Do(req)
	if err != nil {
		return nil, nil, fmt.Errorf("gbizinfo request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized {
		return nil, nil, fmt.Errorf("gbizinfo: 401 Unauthorized (check GBIZINFO_API_TOKEN)")
	}
	if resp.StatusCode != http.StatusOK {
		return nil, nil, fmt.Errorf("gbizinfo: HTTP %d", resp.StatusCode)
	}

	var payload struct {
		Records []GBizRecord `json:"hojin-infos"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, nil, fmt.Errorf("gbizinfo decode: %w", err)
	}

	var nodes []*CompanyNode
	var logs []string
	for _, rec := range payload.Records {
		if rec.CorporateNumber == "" {
			continue
		}
		nodes = append(nodes, &CompanyNode{
			CorporateNumber:  rec.CorporateNumber,
			OfficialName:     rec.Name,
			BusinessCategory: rec.BusinessSummary.MajorClassificationName,
			Address:          rec.Location,
			Website:          rec.CompanyURL,
			MatchScore:       1.0,
			NeedsReview:      false,
		})
		logs = append(logs, fmt.Sprintf("gBizINFO: %s [%s]", rec.Name, rec.CorporateNumber))
	}
	return nodes, logs, nil
}

func fallbackNode(raw *RawCompany) *CompanyNode {
	return &CompanyNode{
		CorporateNumber:      "UNKNOWN_" + NormalizeName(raw.RawName),
		OfficialName:         raw.RawName,
		SourceURLs:           []string{raw.SourceURL},
		Address:              raw.Address,
		Website:              raw.Website,
		Capital:              raw.Capital,
		Employees:            raw.Employees,
		MatchScore:           0,
		NeedsReview:          true,
		RelatedCompaniesText: raw.RelatedCompaniesText,
		BusinessPartnersText: raw.BusinessPartnersText,
	}
}
